File: android/app/src/main/kotlin/com/aispotlight/android/ui/OpenRouterModelBrowser.kt
package com.aispotlight.android.ui

import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalUriHandler
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.aispotlight.android.R
import com.aispotlight.android.core.ModelInfo
import com.aispotlight.android.settings.AppSettings
import java.util.Date
import java.util.Locale

/**
 * The OpenRouter catalog inside the app: search, filters, descriptions and
 * prices from the cached `/models` payload — no trip to the website. A row
 * tap expands the details; "Select" hands the slug back.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun OpenRouterModelBrowserSheet(
    settings: AppSettings,
    onDismiss: () -> Unit,
    onSelect: (String) -> Unit,
) {
    val catalog by settings.openRouterCatalog.collectAsState()
    var query by remember { mutableStateOf("") }
    var filterVision by remember { mutableStateOf(false) }
    var filterFiles by remember { mutableStateOf(false) }
    var filterTools by remember { mutableStateOf(false) }
    var filterReasoning by remember { mutableStateOf(false) }
    var filterFree by remember { mutableStateOf(false) }
    // Hide models the account's privacy settings leave no endpoint for.
    var filterAllowed by remember { mutableStateOf(true) }
    val allowedModels by settings.openRouterAllowedModels.collectAsState()
    var sort by remember { mutableStateOf(0) } // 0 newest, 1 cheapest, 2 name
    var expandedId by remember { mutableStateOf<String?>(null) }
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    val uriHandler = LocalUriHandler.current
    val context = LocalContext.current

    val needle = query.trim().lowercase()
    val models = catalog.values.filter { info ->
        if (filterVision && !info.supportsVision) return@filter false
        if (filterFiles && !info.supportsFiles) return@filter false
        if (filterTools && !info.supportsTools) return@filter false
        if (filterReasoning && !info.supportsReasoning) return@filter false
        if (filterFree && !info.isFree) return@filter false
        if (filterAllowed && allowedModels != null && info.id !in allowedModels!!) return@filter false
        needle.isEmpty() || info.id.lowercase().contains(needle) ||
            (info.name ?: "").lowercase().contains(needle) ||
            (info.summary ?: "").lowercase().contains(needle)
    }.let { list ->
        when (sort) {
            0 -> list.sortedByDescending { it.createdAt ?: 0L }
            1 -> list.sortedBy { (it.promptPricePerToken ?: 0.0) + (it.completionPricePerToken ?: 0.0) }
            else -> list.sortedBy { (it.name ?: it.id).lowercase() }
        }
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = MaterialTheme.colorScheme.surfaceContainer,
    ) {
        Column(Modifier.fillMaxWidth().padding(horizontal = 16.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(stringResource(R.string.or_browser_title), style = MaterialTheme.typography.titleMedium)
                Spacer(Modifier.weight(1f))
                Text(
                    stringResource(R.string.or_count, models.size),
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            Spacer(Modifier.height(8.dp))
            OutlinedTextField(
                value = query,
                onValueChange = { query = it },
                modifier = Modifier.fillMaxWidth(),
                label = { Text(stringResource(R.string.or_search)) },
                singleLine = true,
                colors = eclipseFieldColors(),
            )
            Spacer(Modifier.height(8.dp))
            Row(
                horizontalArrangement = Arrangement.spacedBy(6.dp),
                modifier = Modifier.fillMaxWidth().horizontalScroll(rememberScrollState()),
            ) {
                EclipseChip(stringResource(R.string.or_filter_vision), filterVision, onClick = { filterVision = !filterVision })
                EclipseChip(stringResource(R.string.or_filter_files), filterFiles, onClick = { filterFiles = !filterFiles })
                EclipseChip(stringResource(R.string.or_filter_tools), filterTools, onClick = { filterTools = !filterTools })
                EclipseChip(stringResource(R.string.or_filter_reasoning), filterReasoning, onClick = { filterReasoning = !filterReasoning })
                EclipseChip(stringResource(R.string.or_filter_free), filterFree, onClick = { filterFree = !filterFree })
                if (allowedModels != null) {
                    EclipseChip(stringResource(R.string.or_filter_allowed), filterAllowed, onClick = { filterAllowed = !filterAllowed })
                }
            }
            Spacer(Modifier.height(6.dp))
            Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                EclipseChip(stringResource(R.string.or_sort_newest), sort == 0, onClick = { sort = 0 })
                EclipseChip(stringResource(R.string.or_sort_cheapest), sort == 1, onClick = { sort = 1 })
                EclipseChip(stringResource(R.string.or_sort_name), sort == 2, onClick = { sort = 2 })
            }
            Spacer(Modifier.height(8.dp))
            HorizontalDivider()
            LazyColumn(Modifier.fillMaxWidth().weight(1f, fill = false)) {
                items(models, key = { it.id }) { info ->
                    val expanded = expandedId == info.id
                    val allowed = allowedModels?.contains(info.id) ?: true
                    Column(
                        Modifier
                            .fillMaxWidth()
                            .clickable { expandedId = if (expanded) null else info.id }
                            .padding(vertical = 8.dp)
                            .alpha(if (allowed) 1f else 0.45f),
                    ) {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Text(
                                info.name ?: info.id,
                                style = MaterialTheme.typography.bodyMedium,
                                maxLines = 1,
                                overflow = TextOverflow.Ellipsis,
                                modifier = Modifier.weight(1f),
                            )
                            Spacer(Modifier.width(8.dp))
                            Text(
                                priceLabel(info, context),
                                style = MaterialTheme.typography.labelSmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant,
                            )
                        }
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Text(
                                info.id,
                                style = MaterialTheme.typography.labelSmall.copy(fontFamily = FontFamily.Monospace),
                                color = MaterialTheme.colorScheme.onSurfaceVariant,
                                maxLines = 1,
                                overflow = TextOverflow.Ellipsis,
                                modifier = Modifier.weight(1f),
                            )
                            info.contextLength?.let {
                                Spacer(Modifier.width(8.dp))
                                Text(
                                    compact(it),
                                    style = MaterialTheme.typography.labelSmall,
                                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                                )
                            }
                        }
                        Text(
                            capabilitySummary(info),
                            style = MaterialTheme.typography.labelSmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                        if (!allowed) {
                            Text(
                                stringResource(R.string.or_not_allowed_detail),
                                style = MaterialTheme.typography.labelSmall,
                                color = MaterialTheme.colorScheme.error,
                            )
                        }
                        if (expanded) {
                            Spacer(Modifier.height(6.dp))
                            Text(
                                info.summary?.takeIf { it.isNotBlank() } ?: stringResource(R.string.or_no_description),
                                style = MaterialTheme.typography.bodySmall,
                            )
                            Spacer(Modifier.height(6.dp))
                            val details = buildList {
                                add("${stringResource(R.string.or_per_1m)}: ${priceLabel(info, context)}")
                                info.contextLength?.let { add("${stringResource(R.string.or_context)}: ${compact(it)}") }
                                info.maxCompletionTokens?.let { add("${stringResource(R.string.or_max_output)}: ${compact(it)}") }
                                info.createdAt?.let {
                                    add("${stringResource(R.string.or_added)}: " +
                                        java.text.DateFormat.getDateInstance(java.text.DateFormat.MEDIUM).format(Date(it * 1000)))
                                }
                            }
                            for (line in details) {
                                Text(line, style = MaterialTheme.typography.labelSmall,
                                    color = MaterialTheme.colorScheme.onSurfaceVariant)
                            }
                            Spacer(Modifier.height(6.dp))
                            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                                EclipseTextButton(stringResource(R.string.or_select), onClick = { onSelect(info.id) })
                                EclipseTextButton(stringResource(R.string.or_open_page), onClick = {
                                    uriHandler.openUri("https://openrouter.ai/${info.id}")
                                })
                            }
                        }
                    }
                    HorizontalDivider()
                }
            }
            Spacer(Modifier.height(16.dp))
        }
    }
}

private fun capabilitySummary(info: ModelInfo): String {
    val parts = mutableListOf<String>()
    if (info.supportsVision) parts.add("images")
    if (info.supportsFiles) parts.add("documents")
    if (info.supportsTools) parts.add("tools")
    if (info.supportsReasoning) parts.add("reasoning")
    if (info.isFree) parts.add("free")
    return parts.joinToString(" · ")
}

/** "$2.00 / $10.00" per 1M tokens; "free" when both are zero. */
private fun priceLabel(info: ModelInfo, context: android.content.Context): String {
    if (info.isFree) return context.getString(R.string.or_filter_free).lowercase()
    val input = (info.promptPricePerToken ?: 0.0) * 1_000_000
    val output = (info.completionPricePerToken ?: 0.0) * 1_000_000
    return "${money(input)} / ${money(output)}"
}

private fun money(value: Double): String =
    if (value < 1) String.format(Locale.US, "$%.3f", value) else String.format(Locale.US, "$%.2f", value)

/** 1000000 → "1M", 128000 → "128K". */
private fun compact(tokens: Int): String = when {
    tokens >= 1_000_000 -> {
        val m = tokens / 1_000_000.0
        if (m == Math.floor(m)) "${m.toInt()}M" else String.format(Locale.US, "%.1fM", m)
    }
    tokens >= 1_000 -> "${tokens / 1_000}K"
    else -> tokens.toString()
}

File: Cuate/Views/OpenRouterModelBrowser.swift
import SwiftUI
import AppKit

/// The OpenRouter catalog inside the app: search, filters, descriptions and
/// prices from the cached `/models` payload — no trip to the website. A row
/// click shows the details; "Select" hands the slug to the model field.
struct OpenRouterModelBrowser: View {
    @ObservedObject var settings: AppSettings
    let onSelect: (String) -> Void

    @Environment(\.dismiss) private var dismiss
    @State private var query = ""
    @State private var filterVision = false
    @State private var filterFiles = false
    @State private var filterTools = false
    @State private var filterReasoning = false
    @State private var filterFree = false
    /// Hide models the account's privacy settings leave no endpoint for.
    @State private var filterAllowed = true
    @State private var sort: Sort = .newest
    @State private var selectedID: String?
    @State private var refreshing = false

    enum Sort: String, CaseIterable, Identifiable {
        case newest, cheapest, name
        var id: String { rawValue }
        var label: String {
            switch self {
            case .newest: return L("or.sort.newest")
            case .cheapest: return L("or.sort.cheapest")
            case .name: return L("or.sort.name")
            }
        }
    }

    private var models: [ModelInfo] {
        let needle = query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        var list = settings.openRouterCatalog.values.filter { info in
            if filterVision && !info.supportsVision { return false }
            if filterFiles && info.supportsFiles != true { return false }
            if filterTools && !info.supportsTools { return false }
            if filterReasoning && !info.supportsReasoning { return false }
            if filterFree && !info.isFree { return false }
            if filterAllowed, let allowed = settings.openRouterAllowedModels, !allowed.contains(info.id) { return false }
            guard !needle.isEmpty else { return true }
            return info.id.lowercased().contains(needle)
                || (info.name ?? "").lowercased().contains(needle)
                || (info.summary ?? "").lowercased().contains(needle)
        }
        switch sort {
        case .newest:
            list.sort { ($0.createdAt ?? 0) > ($1.createdAt ?? 0) }
        case .cheapest:
            list.sort { Self.pricePair($0) < Self.pricePair($1) }
        case .name:
            list.sort { ($0.name ?? $0.id).localizedCaseInsensitiveCompare($1.name ?? $1.id) == .orderedAscending }
        }
        return list
    }

    private var selected: ModelInfo? {
        selectedID.flatMap { settings.openRouterCatalog[$0] }
    }

    var body: some View {
        VStack(spacing: 0) {
            header
            Divider()
            HSplitView {
                list
                    .frame(minWidth: 500, idealWidth: 560)
                detail
                    .frame(minWidth: 340)
            }
            Divider()
            footer
        }
        .frame(minWidth: 940, idealWidth: 1020, minHeight: 560, idealHeight: 640)
    }

    // MARK: - Header (search, filters, sort, refresh)

    private var header: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                Text(L("or.browserTitle")).font(.headline)
                Spacer()
                Text(String(format: L("or.count"), models.count))
                    .font(.caption)
                    .foregroundColor(.secondary)
            }
            TextField(L("or.search"), text: $query)
                .textFieldStyle(.roundedBorder)
            HStack(spacing: 6) {
                filterChip(L("or.filter.vision"), $filterVision)
                filterChip(L("or.filter.files"), $filterFiles)
                filterChip(L("or.filter.tools"), $filterTools)
                filterChip(L("or.filter.reasoning"), $filterReasoning)
                filterChip(L("or.filter.free"), $filterFree)
                if settings.openRouterAllowedModels != nil {
                    filterChip(L("or.filter.allowed"), $filterAllowed)
                }
                Spacer(minLength: 0)
            }
            HStack(spacing: 10) {
                Picker("", selection: $sort) {
                    ForEach(Sort.allCases) { Text($0.label).tag($0) }
                }
                .pickerStyle(.segmented)
                .labelsHidden()
                .frame(width: 260)
                Spacer()
                Button {
                    refreshing = true
                    Task {
                        try? await settings.refreshOpenRouterCatalog()
                        refreshing = false
                    }
                } label: {
                    if refreshing {
                        ProgressView().controlSize(.small)
                    } else {
                        Image(systemName: "arrow.clockwise")
                    }
                }
                .disabled(refreshing)
                .help(L("or.refreshHelp"))
            }
        }
        .padding(12)
    }

    private func filterChip(_ label: String, _ on: Binding<Bool>) -> some View {
        Button {
            on.wrappedValue.toggle()
        } label: {
            Text(label)
                .font(.caption)
                .lineLimit(1)
                .fixedSize()
                .padding(.horizontal, 8)
                .padding(.vertical, 3)
                .background(
                    Capsule().fill(on.wrappedValue ? Color.accentColor.opacity(0.25) : Color.secondary.opacity(0.12))
                )
                .overlay(Capsule().stroke(on.wrappedValue ? Color.accentColor : .clear, lineWidth: 1))
        }
        .buttonStyle(.plain)
    }

    // MARK: - List

    private var list: some View {
        List(models, id: \.id, selection: $selectedID) { info in
            VStack(alignment: .leading, spacing: 3) {
                HStack(alignment: .firstTextBaseline, spacing: 8) {
                    Text(info.name ?? info.id)
                        .font(.callout)
                        .lineLimit(1)
                        .truncationMode(.tail)
                    Spacer(minLength: 8)
                    Text(Self.priceLabel(info))
                        .font(.caption.monospacedDigit())
                        .foregroundColor(.secondary)
                        .fixedSize()
                }
                HStack(spacing: 8) {
                    Text(info.id)
                        .font(.system(.caption2, design: .monospaced))
                        .foregroundColor(.secondary)
                        .lineLimit(1)
                        .truncationMode(.tail)
                    Spacer(minLength: 8)
                    if let context = info.contextLength {
                        Text(Self.compact(context))
                            .font(.caption2.monospacedDigit())
                            .foregroundColor(.secondary)
                            .fixedSize()
                    }
                    capabilityIcons(info)
                }
            }
            .padding(.vertical, 3)
            .opacity(isAllowed(info) ? 1 : 0.45)
            .tag(info.id)
        }
        .listStyle(.inset)
    }

    // MARK: - Detail

    @ViewBuilder
    private var detail: some View {
        if let info = selected {
            ScrollView {
                VStack(alignment: .leading, spacing: 10) {
                    Text(info.name ?? info.id).font(.title3.weight(.semibold))
                    HStack(spacing: 6) {
                        Text(info.id)
                            .font(.system(.caption, design: .monospaced))
                            .foregroundColor(.secondary)
                            .textSelection(.enabled)
                        Button {
                            NSPasteboard.general.clearContents()
                            NSPasteboard.general.setString(info.id, forType: .string)
                        } label: {
                            Image(systemName: "doc.on.doc").font(.caption)
                        }
                        .buttonStyle(.plain)
                        .help(L("or.copy"))
                    }
                    HStack(spacing: 6) {
                        if info.supportsVision { chip(L("cap.vision")) }
                        if info.supportsFiles == true { chip(L("cap.documents")) }
                        if info.supportsTools { chip(L("cap.tools")) }
                        if info.supportsReasoning { chip(L("cap.reasoning")) }
                        if info.isFree { chip(L("or.filter.free")) }
                    }
                    if !isAllowed(info) {
                        Text(L("or.notAllowedDetail"))
                            .font(.caption)
                            .foregroundColor(.orange)
                            .fixedSize(horizontal: false, vertical: true)
                    }
                    Text(info.summary?.isEmpty == false ? info.summary! : L("or.noDescription"))
                        .font(.callout)
                        .fixedSize(horizontal: false, vertical: true)
                        .textSelection(.enabled)
                    Grid(alignment: .leading, horizontalSpacing: 12, verticalSpacing: 4) {
                        GridRow {
                            Text(L("or.per1M")).foregroundColor(.secondary)
                            Text(Self.priceLabel(info)).monospacedDigit()
                        }
                        if let context = info.contextLength {
                            GridRow {
                                Text(L("or.context")).foregroundColor(.secondary)
                                Text(Self.compact(context)).monospacedDigit()
                            }
                        }
                        if let maxOut = info.maxCompletionTokens {
                            GridRow {
                                Text(L("or.maxOutput")).foregroundColor(.secondary)
                                Text(Self.compact(maxOut)).monospacedDigit()
                            }
                        }
                        if let created = info.createdAt {
                            GridRow {
                                Text(L("or.added")).foregroundColor(.secondary)
                                Text(Date(timeIntervalSince1970: created), style: .date)
                            }
                        }
                    }
                    .font(.callout)
                    if let url = URL(string: "https://openrouter.ai/\(info.id)") {
                        Link(L("or.openPage"), destination: url).font(.caption)
                    }
                }
                .padding(12)
                .frame(maxWidth: .infinity, alignment: .leading)
            }
        } else {
            VStack {
                Spacer()
                Text(L("or.pickHint"))
                    .font(.callout)
                    .foregroundColor(.secondary)
                    .multilineTextAlignment(.center)
                    .padding()
                Spacer()
            }
            .frame(maxWidth: .infinity)
        }
    }

    // MARK: - Footer

    private var footer: some View {
        HStack {
            Spacer()
            Button(L("local.cancel")) { dismiss() }
                .keyboardShortcut(.cancelAction)
            Button(L("or.select")) {
                if let id = selectedID { onSelect(id) }
            }
            .keyboardShortcut(.defaultAction)
            .disabled(selectedID == nil)
        }
        .padding(12)
    }

    // MARK: - Helpers

    /// Unknown (no key) counts as allowed — nothing to grey out.
    private func isAllowed(_ info: ModelInfo) -> Bool {
        settings.openRouterAllowedModels?.contains(info.id) ?? true
    }

    private func chip(_ label: String) -> some View {
        Text(label)
            .font(.caption2)
            .lineLimit(1)
            .fixedSize()
            .padding(.horizontal, 6)
            .padding(.vertical, 2)
            .background(Color.secondary.opacity(0.15), in: Capsule())
            .foregroundColor(.secondary)
    }

    /// Capabilities as small glyphs — they never wrap; the words live in the
    /// detail pane and in each glyph's tooltip.
    private func capabilityIcons(_ info: ModelInfo) -> some View {
        HStack(spacing: 5) {
            if info.supportsVision { capabilityIcon("photo", L("cap.vision")) }
            if info.supportsFiles == true { capabilityIcon("doc.text", L("cap.documents")) }
            if info.supportsTools { capabilityIcon("wrench.and.screwdriver", L("cap.tools")) }
            if info.supportsReasoning { capabilityIcon("brain", L("cap.reasoning")) }
            if info.isFree { capabilityIcon("gift", L("or.filter.free")) }
        }
        .fixedSize()
    }

    private func capabilityIcon(_ symbol: String, _ label: String) -> some View {
        Image(systemName: symbol)
            .font(.caption2)
            .foregroundColor(.secondary)
            .frame(width: 14)
            .help(label)
    }

    /// "$2.00 / $10.00" per 1M tokens; "free" when both are zero.
    static func priceLabel(_ info: ModelInfo) -> String {
        if info.isFree { return L("or.filter.free").lowercased() }
        let input = (info.promptPricePerToken ?? 0) * 1_000_000
        let output = (info.completionPricePerToken ?? 0) * 1_000_000
        return "\(Self.money(input)) / \(Self.money(output))"
    }

    private static func money(_ value: Double) -> String {
        value < 1 ? String(format: "$%.3f", value) : String(format: "$%.2f", value)
    }

    private static func pricePair(_ info: ModelInfo) -> Double {
        (info.promptPricePerToken ?? 0) + (info.completionPricePerToken ?? 0)
    }

    /// 1000000 → "1M", 128000 → "128K".
    static func compact(_ tokens: Int) -> String {
        if tokens >= 1_000_000 {
            let m = Double(tokens) / 1_000_000
            return m == m.rounded() ? "\(Int(m))M" : String(format: "%.1fM", m)
        }
        if tokens >= 1_000 { return "\(tokens / 1_000)K" }
        return "\(tokens)"
    }
}
